### PR TITLE
A quick opinionated review

### PR DESCRIPTION
**Pull request by malob**
I didn't try to hard to understand the code, but mostly just did a quick pass to, first reorganizing the project, then a bunch of opinionated style edits/refactors (I didn't look a the tests much).

I'll leave some specific review comments.

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -30,36 +30,43 @@ import Retransmit (retransmit)
...
+                "Addresses of other servers to forward to.  Repeat for multiple peers."
+          )
+      )
+    <*> option
+      auto
+      ( long "port"
+          <> value defaultPort
+          <> showDefault
+          <> help
+            "TCP port.  This is both the port on which this instance will listen and the default port for connecting to peers with unspecified ports."
```
Comment (malob) [inline app/Main.hs:68]: Reformatted to make it easier to follow. I generally like trying to line up operators, and space things away from parens when they span multiple lines.

File: app/Main.hs [hunk 2]
```diff
@@ -71,9 +78,9 @@ main :: IO ()
 main = do
   args <- execParser parserInfo
   password <- Pass <$> BS.readFile (passwordFile args)
-  localBS <- newBlobDir (blobdir args)
+  localBS <- newBlobDir $ blobdir args
```
Comment (malob) [inline app/Main.hs:81]: Super nitpick. But I prefer dollars to parens in most cases.

File: app/Main.hs [hunk 3]
```diff
@@ -71,9 +78,9 @@ main :: IO ()
...
   peerClients <-
-    mapM (asyncRetryingTCPClient (blobdir args) (port args)) (peers args)
+    traverse (asyncRetryingTCPClient (blobdir args) (port args)) (peers args)
```
Comment (malob) [inline app/Main.hs:83]: `mapM` should really be removed from `Monad`, now that the `Traversable` type class exists (where `traverse` it only has an `Applicative` constraint), but it's still there for compatibility sake.

File: app/MainSend.hs [hunk 4]
```diff
@@ -76,31 +80,34 @@ parserInfo =
     (fullDesc <> header "bejnarkli-send - Send a blob to bejnarkli servers")
 
 validateArgs ::
-     ParserPrefs -> ParserInfo a -> ParserResult Args -> ParserResult Args
```
Comment (malob) [inline app/MainSend.hs:79]: Looks like I missed moving the `::` to the next line here. Oops!

File: app/Main.hs (hunk 1 above)
Comment (malob) [reply app/Main.hs:68]: You'll see this a lot in my changes, so I won't comment on it every time.

File: app/MainSend.hs [hunk 5]
```diff
@@ -76,31 +80,34 @@ parserInfo =
...
 parse :: ParserInfo Args -> IO Args
-parse pinfo =
-  validateArgs defaultPrefs pinfo . execParserPure defaultPrefs pinfo <$>
-  getArgs >>=
-  handleParseResult
+parse pinfo
+  = handleParseResult
+  . validateArgs defaultPrefs pinfo
+  . execParserPure defaultPrefs pinfo
+  =<< getArgs
```
Comment (malob) [inline app/MainSend.hs:101]: The `<$>` here was redundant. Also lined up the operators to make the composition pipeline easier to follow. That's also why I used `=<<` rather than `>>=`, so that everything flows in the same direction.

File: app/MainSend.hs [hunk 6]
```diff
@@ -76,31 +80,34 @@ parserInfo =
...
   hash <- runConduitRes $ sourceFile (file args) .| blobName password
   atLeast_ (fromMaybe (length (servers args)) (durability args)) $
-    map
-      (\server ->
-         retryingTCPClient defaultPort server hash (sourceFile (file args)))
+    fmap
+      ( \server ->
+          retryingTCPClient defaultPort server hash (sourceFile (file args))
+      )
       (servers args)
```
Comment (malob) [inline app/MainSend.hs:113]: Never use `map`.

File: hie.yaml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+      component: "bejnarkli:test:persistent-once-test"
+
+    - path: "test"
+      component: "bejnarkli:test:queue-test"
+
+    - path: "test"
+      component: "bejnarkli:test:retry-queue-test"
+
+    - path: "test"
+      component: "bejnarkli:test:replicatedblobstore-test"
```
Comment (malob) [inline hie.yaml:28]: This file is generated with the `gen-hie` command from `implicit-hie`. It tells the language server about all the projects components. HLS will have this build in a some point, but it doesn't right now.

File: shell.nix [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+pkgs.mkShell {
+  buildInputs = with pkgs; [
+    ghc
+    cabal-install
+    zlib
+
+    # keep this line if you use bash
+    bashInteractive
+  ];
+}
```
Comment (malob) [inline shell.nix:12]: Added a `shell.nix` file so it's easy to setup an env, with the required tools/packages. In my case I didn't have GHC or Cabal installed, and the project wouldn't build without `zlib`. This should probably specify the GHC version explicitly, and pin to a `nixpkgs` commit, but I was lazy, and just wanted to get up and running.

File: src/Async.hs [hunk 9]
```diff
@@ -13,7 +15,7 @@ removeOne p (x:xs)
   | otherwise = (x :) <$> removeOne p xs
 
 waitAtLeast :: Int -> [Async a] -> IO [a]
-waitAtLeast 0 actions = mapM_ cancel actions >> pure []
+waitAtLeast 0 actions = traverse_ cancel actions $> []
```
Comment (malob) [inline src/Async.hs:18]: `>>` is another example of a function that is still in `Monad` for compatibility reasons. `*>` from `Applicative` should be used instead. In this case we can do even better with `$>` from `Functor` which is really since it allows us to do way with the `pure`.

File: src/Bejnarkli.hs [hunk 10]
```diff
@@ -30,8 +29,8 @@ protocolVersion = fromIntegral $ ord 'B'
 defaultPort :: Int
 defaultPort = 8934
 
-bejnarkliServer ::
-     (BlobStore blobstore, MonadResource m)
+bejnarkliServer
+  :: (BlobStore blobstore, MonadResource m)
```
Comment (malob) [inline src/Bejnarkli.hs:33]: Lining up operators again, this time in the type signature.

File: src/Bejnarkli.hs [hunk 11]
```diff
@@ -46,25 +45,21 @@ bejnarkliServer bs password = do
     else yield $ U8S.fromString "n"
 
 idC :: Monad m => ConduitT i i m ()
-idC = await >>= maybe (return ()) (\a -> yield a >> idC)
+idC = await >>= maybe (pure ()) (\a -> yield a *> idC)
```
Comment (malob) [inline src/Bejnarkli.hs:48]: Never `return`, always `pure` from `Applicative`.

File: src/Bejnarkli.hs [hunk 12]
```diff
@@ -46,25 +45,21 @@ bejnarkliServer bs password = do
...
 
 prependSource :: Monad m => i -> ConduitT i i m ()
 prependSource prefix =
   getZipConduit (ZipConduit (yield prefix) *> ZipConduit idC)
 
 parseResponse :: Monad m => ConduitT BS.ByteString Void m Bool
-parseResponse = do
-  response <- await
-  pure $ response == Just (U8S.fromString "y")
+parseResponse = (== Just (U8S.fromString "y")) <$> await
```
Comment (malob) [inline src/Bejnarkli.hs:55]: I don't like giving names to things, unless it really helps with readability. I also like to avoid `do` notation unless it make the code **much** more ergonomic to write or **much** easier to understand.

File: src/BlobStore.hs [hunk 13]
```diff
@@ -56,62 +59,54 @@ import Test.QuickCheck.Instances.ByteString ()
...
 
-blobName ::
-     MonadResource m => Password -> ConduitT BS.ByteString o m BS.ByteString
-blobName (Pass password) = do
-  digest <- sinkHMAC password
-  pure $ BA.convert (digest :: HMAC Crypto.Hash.Algorithms.SHA256)
+blobName
+  :: MonadResource m
+  => Password -> ConduitT BS.ByteString o m BS.ByteString
+blobName (Pass password) = BA.convert @(HMAC SHA256) <$> sinkHMAC password
```
Comment (malob) [inline src/BlobStore.hs:85]: `TypeApplications` is really nice. I strongly advocate using it instead of inline type annotations. In this case you'll notice that `BA.convert` is a function, and if I'd used a type annotation, I would have had to provide the type for the function. `TypeApplications` like function application, but for types, so you just provide the type for each type variable in the order they appear.

File: src/BlobStore.hs [hunk 14]
```diff
@@ -56,62 +59,54 @@ import Test.QuickCheck.Instances.ByteString ()
...
   getBlob :: a -> ExtantBlobName -> ConduitT i BS.ByteString (ResourceT IO) ()
 
-sinkNamePrefixedBlob ::
-     (BlobStore bs, MonadResource m)
-  => bs
-  -> Password
-  -> ConduitT BS.ByteString o m (Maybe ExtantBlobName)
+sinkNamePrefixedBlob
+  :: (BlobStore bs, MonadResource m)
+  => bs -> Password -> ConduitT BS.ByteString o m (Maybe ExtantBlobName)
```
Comment (malob) [inline src/BlobStore.hs:101]: When breaking a functions type signature onto multiple lines, I like to keep the constrains and types of the arguments together, unless they either hit my line length limit.

File: src/BlobStore.hs [hunk 15]
```diff
@@ -56,62 +59,54 @@ import Test.QuickCheck.Instances.ByteString ()
...
 sinkNamePrefixedBlob bs password = do
   name <- takeExactlyE blobNameLength await
   case name of
-    Just aname
-      | BS.length aname == blobNameLength -> sinkUntrustedBlob bs password aname
+    Just aname | BS.length aname == blobNameLength ->
+      sinkUntrustedBlob bs password aname
```
Comment (malob) [inline src/BlobStore.hs:106]: I think it's easier to read here with the pattern match and condition on the same line.

File: src/BlobStore.hs [hunk 16]
```diff
@@ -133,13 +128,11 @@ instance BlobStore BlobMapStore where
...
-newtype BlobDirStore =
-  BlobDir FilePath
+newtype BlobDirStore = BlobDir FilePath
 
 newBlobDir :: FilePath -> IO BlobDirStore
-newBlobDir path = do
-  createDirectoryIfMissing True (path </> ".incoming")
-  pure $ BlobDir path
+newBlobDir path =
+  createDirectoryIfMissing True (path </> ".incoming") $> BlobDir path
```
Comment (malob) [inline src/BlobStore.hs:135]: No `do` notation needed.

File: src/BlobStore.hs [hunk 17]
```diff
@@ -161,30 +154,27 @@ instance BlobStore BlobDirStore where
   getBlob bd (ExtantBlob name) = sourceFile (blobFileName bd name)
 
 blobFileName :: BlobDirStore -> BS.ByteString -> FilePath
-blobFileName (BlobDir d) blobname = d </> toString (Base64.encode blobname)
+blobFileName (BlobDir d) = (d </>) . toString . Base64.encode
```
Comment (malob) [inline src/BlobStore.hs:157]: Definitely a nit, but I like to keep things point-free unless it makes things confusing.

File: src/BlobStore.hs [hunk 18]
```diff
@@ -161,30 +154,27 @@ instance BlobStore BlobDirStore where
...
-        then Just n
-        else Nothing
-
-sinkTrustedBlob ::
-     (BlobStore bs, MonadResource m)
+unBlobFileName
+  = find ((== blobNameLength) . BS.length)
+  . either (const Nothing) Just
+  . Base64.decode
+  . fromString
```
Comment (malob) [inline src/BlobStore.hs:164]: When if first looked at this, the `Left _ -> Nothing` got my attention. I personally get a code smell when I see  simple pattern matches that are dropping their values, as well as simple cases on `Either` or `Maybe` values. This got me thinking about maybe using the `either` combinator. Additionally when using `if` to eventually return a `Maybe`, if often think using `find :: Foldable t => (a -> Bool) -> t a -> Maybe a`. 

Given that I like my composition pipelines, I like my refactor. But overall I think this is mostly a nit and personal preference.

File: src/BlobStore.hs [hunk 19]
```diff
@@ -161,30 +154,27 @@ instance BlobStore BlobDirStore where
...
   -> ConduitT BS.ByteString o m ExtantBlobName
-sinkTrustedBlob bs password = do
-  (staged, name) <-
-    getZipConduit
-      (ZipConduit ((,) <$> sinkBlob bs) <*> ZipConduit (blobName password))
-  liftIO $ commit staged name
+sinkTrustedBlob bs password =
+  getZipConduit
+    (ZipConduit ((,) <$> sinkBlob bs) <*> ZipConduit (blobName password))
+    >>= liftIO . uncurry commit
```
Comment (malob) [inline src/BlobStore.hs:174]: `uncurry` is your friend. Get that `do` out of there I say!

File: src/Queue.hs [hunk 20]
```diff
@@ -16,6 +17,7 @@ enqueue :: a -> Queue a -> Queue a
 enqueue x (Q ins outs) = Q (x : ins) outs
 
 dequeue :: Queue a -> (Maybe a, Queue a)
-dequeue q@(Q [] []) = (Nothing, q)
-dequeue (Q ins (out:outs)) = (Just out, Q ins outs)
-dequeue (Q ins []) = dequeue (Q [] (reverse ins))
+dequeue = \case
+  q@(Q [] []) -> (Nothing, q)
+  Q ins (out:outs) -> (Just out, Q ins outs)
+  Q ins [] -> dequeue (Q [] (reverse ins))
```
Comment (malob) [inline src/Queue.hs:23]: Very much a nit, but when doing a pattern match on a function with one argument, I think `\case` is cleaner, and also has the nice property of not needed to rewrite the function name on every line.

File: src/Retry.hs [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+
+  where
+    process :: Float -> IO ()
+    process prevBackoff = (== Right True) <$> try @IOError action >>= \case
+      True -> pure ()
+      False -> do
+        let backoff = nextBackoff params False prevBackoff
+        delay <- getStdRandom (randomR (0, 2 * backoff))
+        threadDelay $ round $ 100000 * delay
+        process backoff
```
Comment (malob) [inline src/Retry.hs:44]: This was the original code:

```haskell
    process prevBackoff = do
       success <- (== Right True) <$> (try action :: IO (Either IOError Bool))
       let backoff = nextBackoff params success prevBackoff
        in if success
             then pure ()
             else do
               delay <- getStdRandom (randomR (0, 2 * backoff))
               threadDelay $ round $ 100000 * delay
               process backoff
```

First, just a quick note that in a `do` block you don't need to include `in` here after a `let`.

I noticed that `backoff` only needs to be defined in the `else` branch here, and it seems better to just put it there. Also, seemed weird to me to use `success` in the definition of `backoff` since `backoff` is only used when `success` is `False`.

 The rest of my refactor here was motivated by my general policy of avoiding `do`, and preference for using case statements rather that if statements.

File: src/Retry.hs [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    process prevBackoff chan = do
+      item <- readChan chan
+      (== Right True) <$> try @IOError (f item) >>= \case
+        True -> process (minDelay params) chan
+        False -> do
+          let backoff = nextBackoff params False prevBackoff
+          delay <- getStdRandom (randomR (0, 2 * backoff))
+          threadDelay $ round $ 100000 * delay
+          writeChan chan item
+          process backoff chan
```
Comment (malob) [inline src/Retry.hs:68]: Orignial code:

```haskell
retryQueue params f =
   newChan >>=
   (\chan ->
      let process prevBackoff = do
            item <- readChan chan
            success <-
              (== Right True) <$> (try $ f item :: IO (Either IOError Bool))
            let backoff = nextBackoff params success prevBackoff
             in if success
                  then process (minDelay params) :: IO ()
                  else do
                    delay <- getStdRandom (randomR (0, 2 * backoff))
                    threadDelay $ round $ 100000 * delay
                    writeChan chan item
                    process backoff :: IO ()
       in do _ <- forkIO $ process (minDelay params) -- TODO: Allow thread clean-up
             pure $ writeChan chan)
```

See I sometimes think `do` notation is the right call :p

In general, I was finding this code quite hard to follow. In staring at it, I noticed it was quite similar to the function above, which gave me a hint at how I might want to refactor it. 

Also, notice the sweat sweat use of `TypeApplications`.

File: src/TCPClient.hs [hunk 23]
```diff
@@ -46,26 +49,21 @@ import Retry
...
+  uri <- find validURI (parseURI $ "bejnarkli://" ++ name)
+  auth <- find (null . uriUserInfo) $ uriAuthority uri
+  port <- readMaybe =<< stripPrefix ":" ( uriPort auth)
+  pure (port, uriRegName auth)
+
+  where
+    validURI uri = uriScheme uri == "bejnarkli:"
+      && null (uriPath uri)
+      && null (uriQuery uri)
+      && null (uriFragment uri)
```
Comment (malob) [inline src/TCPClient.hs:62]: I definitely get a code smell when I see a bunch of nested case statement like this, especially when in this case it's a pyramid where all the `Nothing` cases have the same value.

I think this refactor makes the logic much easier to follow. Also note the use of `find` and `null :: Foldable t => t a -> Bool`. Gotta love those general purpose `Foldable` functions.

Comment (malob) [reply src/TCPClient.hs:62]: This is probably my favorite refactor of my review.

File: src/TCPClient.hs [hunk 24]
```diff
@@ -75,29 +73,31 @@ tCPClient defaultPort hostString blobHash blobData =
...
+    app appdata
+      = runConduitRes
+      $ blobData
+     .| bejnarkliClient
+          ( do
+              appSink appdata
+              liftIO (shutdown (fromJust (appRawSocket appdata)) ShutdownSend)
+              appSource appdata
+          )
+          blobHash
```
Comment (malob) [inline src/TCPClient.hs:85]: I do love lining up those operators don't I. I do think it makes the flow here easier to follow.

**Comment by chkno** [comment]
Thank you for the detailed review, especially the opinionated  bits.

I do really, really want to keep formatting 100% automated.  To keep your formatting suggestions from simply being reverted by `hformat`, I switched the formatting automation to `floskell`, which is configurable, and encoded your recommendations as best I could into [a floskell config](https://github.com/chkno/bejnarkli/commit/1254082ac697fb3cd8611a6ffd6a47e9f545598f#diff-7165ff12e8868f044574429e5ad3b8b7f811dd028a7a1d9f89aa2b4d660b4c38).
